sdk/src/types/block/output/output_id_proof.rs

```diff
@@ -1,90 +1,304 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use alloc::{boxed::Box, string::String};
+use alloc::boxed::Box;
+use core::convert::Infallible;
 
-#[cfg(feature = "serde")]
-use {crate::utils::serde::prefix_hex_bytes, alloc::format, serde::de::Deserialize, serde_json::Value};
+use crypto::hashes::{blake2b::Blake2b256, Digest};
+use packable::{Packable, PackableExt};
 
-use crate::types::block::slot::SlotIndex;
+#[cfg(feature = "serde")]
+use crate::utils::serde::prefix_hex_bytes;
+use crate::{
+    types::block::{output::Output, slot::SlotIndex, Error},
+    utils::merkle_hasher::{largest_power_of_two, LEAF_HASH_PREFIX, NODE_HASH_PREFIX},
+};
 
 /// The proof of the output identifier.
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq, Packable)]
 #[cfg_attr(
     feature = "serde",
     derive(serde::Serialize, serde::Deserialize),
     serde(rename_all = "camelCase")
 )]
+#[packable(unpack_error = Error)]
 pub struct OutputIdProof {
     pub slot: SlotIndex,
     pub output_index: u16,
-    pub transaction_commitment: String,
+    #[cfg_attr(feature = "serde", serde(with = "prefix_hex_bytes"))]
+    pub transaction_commitment: [u8; 32],
+    #[packable(verify_with = verify_output_commitment_type)]
     pub output_commitment_proof: OutputCommitmentProof,
 }
 
-#[derive(Clone, Debug, Eq, PartialEq)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize), serde(untagged))]
-pub enum OutputCommitmentProof {
-    HashableNode(HashableNode),
-    LeafHash(LeafHash),
-    ValueHash(ValueHash),
+impl OutputCommitmentProof {
+    pub fn new(outputs: &[Output], index: u16) -> Self {
+        if let [output] = outputs {
+            Self::value(&output.pack_to_vec())
+        } else {
+            let num_outputs = outputs.len() as u16;
+            debug_assert!(num_outputs > 0 && index < num_outputs, "n={num_outputs}, index={index}");
+
+            // Select a `pivot` element to split `data` into two slices `left` and `right`.
+            let pivot = largest_power_of_two(num_outputs as _) as u16;
+            let (left, right) = outputs.split_at(pivot as _);
+
+            if index < pivot {
+                // `value` is contained in the left subtree, and the `right` subtree can be hashed together.
+                Self::node(Self::new(left, index), Self::hash(right))
+            } else {
+                // `value` is contained in the right subtree, and the `left` subtree can be hashed together.
+                Self::node(Self::hash(left), Self::new(right, index - pivot))
+            }
+        }
+    }
+
+    /// Get the merkle tree hash for a list of outputs
+    fn hash(outputs: &[Output]) -> LeafHash {
+        match outputs {
+            [] => LeafHash::empty(),
+            [output] => LeafHash::new(&output.pack_to_vec()),
```

Comment:
Shouldn't this be a `ValueHash`?
`If Leaves Len == 1 return a Value Hash with its Hash field set to Hash Leaf(Leaves[0]).`

Reply:
well this is how iota.go does it

Reply:
Wrong function. That correlates to the `OutputCommitmentProof::new` method. This is the equivalent of `OutputCommitmentProof::hash`: https://github.com/iotaledger/iota.go/blob/70d23c10cabe98a277705d9a60682561ad259664/merklehasher/merkle_hasher.go#L65

Reply:
Thanks